### PR TITLE
Restructure the elm app

### PR DESCRIPTION
**Pull request by erkannt**
Issues this solves:

- the `recommendPathway` logic was growing to a substantial ammount, making the `Main` a scrollfest
- `Types` was becoming unwieldy, in part due to the large amount of `Backend` types required
- `(..)` in most `exposing` and `imports` made it hard to understand dependencies

## Approach

This restructure united all code into a single Main and broke it up again. The principles are as follows:

- break up by _domain_ not _function_ i.e. by paper type not by `view`, `update`
  this leads to:
  - files get created for types of data
  - views live with the data they render
  - only the top level constructor for that data structure gets exposed
- things used by multiple modules move out of `Main` to avoid dependency loops
- no `Msg` sub types, instead a share `Msg.elm` outside of elm

Took me a while to figure out the `Msg` bit. While there are patterns to use `Html.map` so that submodules can have their own `Msg`, this appears to have fallen out of favour due the amount of wiring required.

## References re: `Msg` in submodules

The references that I ended up following are:

- [TodoMVC structured without map along domain](https://github.com/rogeriochaves/structured-elm-todomvc/tree/nomap-domain)
- [Keeping it simple beyond a basic app](https://dev.to/elmupdate/keeping-it-simple-beyond-a-basic-elm-app-569m)

The first here actually splits up the `Msg` per Domain and joins them up in a central `Msg` that can be imported by all.

So if a view wants to emit a `Msg` you end up with stuff like:

```
onClick (Msg.MsgForFreePathwayPaper <| FreePathwayPaperMsg.ToggleVisible id
```
and
```
|> HttpBuilder.withExpect (Http.expectJson (Msg.MsgForMain << MainMsg.GotPaper) paperDecoder) 
```

This necessitates nested `case` statetement in the `Main` `update`.

Doing this (see 762e896 ) allowed me to move the `FreePathwayPaper` view functions to the respective file without resorting to `Html.map`.

Subsequently I understood the approach outlined in the [Keeping it simple](https://dev.to/elmupdate/keeping-it-simple-beyond-a-basic-elm-app-569m) article and moved all `Msg` into a single file (we only have three...)

## Exposing type variates

Syntax I was previously unaware of:

```
module Msg exposing (Msg(..))

...

type Msg
    = GotPaper (Result Http.Error BackendPaper)
    | Animate Animation.Msg
    | ToggleVisible Int
```

The `Msg(..)` exposes the variates. Without this you can't access the specific message say in a `onclick`. It also placates `elm-analyze` regarding the unused constructor.

### CODE REVIEW

**Comment by LGro** [comment]
Starting to dig in and it looks really valuable already, thanks @erkannt!

When I got excited about the paper type specific views now being reachable via `Papers.OpenAccess.view` I wondered why don't we do the same for `Papers.OpenAccess.Paper` instead of the current `Papers.OpenAccess.OpenAccessPaper`? We could then still in other places import `OpenAccess` such that we can do `OpenAccess.Paper` instead of `OpenAccessPaper`, or even alias on import if we mind the dot.

File: elm_frontend/src/Papers/OtherPathway.elm [hunk 1]
```diff
@@ -0,0 +1,44 @@
+module Papers.OtherPathway exposing (OtherPathwayPaper, view, viewList)
+
+import GeneralTypes exposing (PaperMetadata, renderPaperMetaData)
+import Html exposing (Html, div, h2, p, section, text)
+import Html.Attributes exposing (class)
+
+
+
+-- TYPES
```
Comment (LGro) [inline elm_frontend/src/Papers/OtherPathway.elm:9]: We have these types comments in some files like this one where there is only one type declaration and view things below it. How do you feel about removing the comments?

File: elm_frontend/src/GeneralTypes.elm [hunk 2]
```diff
@@ -0,0 +1,58 @@
+module GeneralTypes exposing (DOI, NamedUrl, PaperMetadata, renderPaperMetaData, renderUrl)
```
Comment (LGro) [inline elm_frontend/src/GeneralTypes.elm:1]: The contents of this seems to be general as in it's not part of any individual paper but also still exclusive to papers, it seems. Also, none of the other files with similar contents (types and views) has types in its name, which might be confusing some into believing that this file only contains types.
How about we make this `Papers.General` or `Papers.Utils` or if we find it to fit on the top level `PaperUtils`?

**Comment by erkannt** [comment]
Thanks! Glad you like it.

I'm going to rebase and merge as I find some of the commits useful as pattern reference and branches make me itchy.

Feel free to commit as you see fit. I like the idea of doing something like:

```
import Papers.OpenAccess as OpenAccess
```

I'd suggest we keep the section comments on the larger files.

File: elm_frontend/src/GeneralTypes.elm (hunk 2 above)
Comment (erkannt) [reply elm_frontend/src/GeneralTypes.elm:1]: I like `Papers.Utils`

**Pull request merged**